validator.py

```diff
@@ -1543,6 +1548,104 @@ def validate_evac_ids(self):
                         self.invalid_values += 1
 
 
+    def validate_events(self):
+        ''' 
+        Validates events in every state:
+        1. Source is recommended
+        2. Source and Object must be valid character ids or an EntityTypeEnum
+        '''
+        data = copy.deepcopy(self.loaded_yaml)
+        entity_type_enum = ['ally', 'adversary', 'civilian', 'commander', 'everybody', 'medic', 'tbd']
+        for scene in data['scenes']:
+            events = scene.get('state', {}).get('events', [])
+            chars = self.get_characters_in_scene(data, scene['id'])
+            missing = 0
+            for event in events:
+                source = event.get('source', None)
+                obj = event.get('object', None)
+                if source is None:
+                    missing += 1
+                elif source not in entity_type_enum:
+                    in_any_group = False
+                    in_all_groups = True
+                    for group in chars['possible']:
+                        if source in group:
+                            in_any_group = True
+                        else:
+                            in_all_groups = False
+                    if not in_any_group:
+                        self.logger.log(LogLevel.ERROR, f"The 'source' parameter for an event in scene '{scene['id']}' is '{source}', but must be one of {entity_type_enum + chars['possible']}.")
```

Comment:
I believe it's correct that in this case, we are 100% sure that these are the available characters?

Reply:
We are 100% sure that they might be available to that scene. They may only be available in some branches, but I can make it so it only lists the ones available in all branches, if you'd like

Reply:
All I was saying is that if we're saying "but must be one of [...]" then we should only list ones that it can 100% be.  Otherwise, it could just say generically, `"but must be a character in the scene or one of {entity_type_enum}."`  In other words, you don't have to enumerate them if you're not sure of the complete list.